File: Simple_NN_library/C++/development_version/src/graphing/graphing.cpp
#include <graphing/graphing.h>

using namespace Simple_NN_library;
using namespace graphing;
using namespace std;

Graph::Graph(data_struct* data, string filepath){
  file_path=filepath;
  string file;
  ofstream f;
  ifstream template_;

  f.open(filepath);
  template_.open("template.html");
  string template_data;
  template_>>template_data;
  f<<template_data;
  template_.close();
  f.close();

  update_graph(data);
}
bool Graph::update_graph(data_struct* data){
  ifstream f;
  f.open(file_path);
  string file;
  f>>file;
  f.close();

  size_t pos=file.find("VAR: project_name")-sizeof("VAR: project_name");
  file.replace(pos, sizeof("VAR: project_name"), data.project_name);

  pos=file.find("VAR: epochs")-sizeof("VAR: epochs");
  file.replace(pos, sizeof("VAR: epochs"), data.epochs);

  pos=file.find("VAR: input")-sizeof("VAR: input");
  file.replace(pos, sizeof("VAR: input"), data.input);

  pos=file.find("VAR: output")-sizeof("VAR: output");
  file.replace(pos, sizeof("VAR: output"), data.output);

  ofstream ofs;
  ofs.open(file_path, std::ofstream::out | std::ofstream::trunc);
  ofs.close();

  ofstream f;
  f.open(file_path);
  f<<file;
  f.close();
  return true;
}
void Graph::view_graph(string browser){
  system(browser+"file://"+swap_backslashes(file_path));
}
string Graph::swap_backslashes(string filepath){
  string result;
  for(int i=0; i<filepath.length(); i++){
    if filepath[i]=="\\"&&filepath[i+1]=="\\"{result+="/";}
    else{result+=filepath[i];}
  }
  return result;
}

static bool Graph::update_graph(data_struct* data, string file_path){
  ifstream f;
  f.open(file_path);
  string file;
  f>>file;
  f.close();

  size_t pos=file.find("VAR: project_name")-sizeof("VAR: project_name");
  file.replace(pos, sizeof("VAR: project_name"), data.project_name);

  pos=file.find("VAR: epochs")-sizeof("VAR: epochs");
  file.replace(pos, sizeof("VAR: epochs"), data.epochs);

  pos=file.find("VAR: input")-sizeof("VAR: input");
  file.replace(pos, sizeof("VAR: input"), data.input);

  pos=file.find("VAR: output")-sizeof("VAR: output");
  file.replace(pos, sizeof("VAR: output"), data.output);

  ofstream ofs;
  ofs.open(file_path, ofstream::out|ofstream::trunc);
  ofs.close();

  ofstream f;
  f.open(file_path);
  f<<file;
  f.close();
  return true;
}
static void Graph::view_graph(string file_path, string browser){
  system(browser+"file://"+swap_backslashes(file_path));
}
static string Graph::swap_backslashes(string file_path){
  string result;
  for(int i=0; i<filepath.length(); i++){
    if filepath[i]=="\\"&&filepath[i+1]=="\\"{result+="/";}
    else{result+=filepath[i];}
  }
  return result;
}

File: Simple_NN_library/python/graphing/graphing.py
import webbrowser

class Graph:
    #Data is a dictionary
    def __init__(self, data, file_path):
        self.file_path=file_path

        f=open(self.file_path, 'w')

        template=open('template.html', 'r')
        f.write(template.readlines())
        template.close()
        f.close()

        self.update_graph(data)

    def update_graph(self, data):
        try:
            f=open(self.file_path, 'r+')
            file=f.readlines()
        except Exception as E:
            print(str(e))
            return False

        file.replace("VAR: project_name", data['project_name'])
        file.replace("VAR: epochs", data['epochs'])
        file.replace("VAR: input", data['input'])
        file.replace("VAR: output", data['output'])

        f.seek(0)
        f.truncate()

        f.write(file)
        f.close()
        return True

    def view_graph(self):
        return webbrowser.open('file://'+str(self.swap_backslashes(self.file_path)))

    def swap_backslashes(self, file_path):
        result=""
        for i in range(len(file_path)):
            if file_path[i:i+1]=="\\":result+="/"
            else:result+=file_path[i]
        return result

    @staticmethod
    def update_graph(self, data, file_path):
        try:
            f=open(file_path, 'r+')
            file=f.readlines()
        except Exception as E:
            print(str(e))
            return False

        file.replace("VAR: project_name", data['project_name'])
        file.replace("VAR: epochs", data['epochs'])
        file.replace("VAR: input", data['input'])
        file.replace("VAR: output", data['output'])

        f.seek(0)
        f.truncate()

        f.write(file)
        f.close()
        return True


    @staticmethod
    def view_graph(self, file_path):
        return webbrowser.open('file://'+str(Graph.swap_backslashes(file_path)))

    @staticmethod
    def swap_backslashes(self, file_path):
        result=""
        for i in range(len(file_path)):
            if file_path[i:i+1]=="\\":result+="/"
            else:result+=file_path[i]
        return result
